app/services/users/inactivity_notification_and_deletion_service.rb

```diff
@@ -0,0 +1,82 @@
+module Users
+  class InactivityNotificationAndDeletionService
+    include RoutesResolvable
+
+    def initialize(debug: false)
+      @debug = debug
+      @debug_value = {}
+    end
+
+    def execute
+      notify_users_of_deletion
+      delete_inactive_users
+      @debug_value
+    end
+
+    private
+
+    def notify_users_of_deletion
+      if @debug
+        @debug_value[:users_to_notify] = users_to_notify_deletion.map(&:id)
+      else
+        users_to_notify_deletion.each do |user|
+          UserMailer.notify_account_deletion(user, login_url(user.school), inactivity_duration_for_user_deletion(user) - 1).deliver_later
+          user.update!(account_expiry_notification_sent_at: Time.zone.now)
+        end
+      end
+    end
+
+    def delete_inactive_users
+      if @debug
+        @debug_value[:users_to_delete] = users_to_delete.map(&:id)
+      else
+        users_to_delete.each do |user|
+          Users::DeleteAccountJob.perform_later(user)
+        end
+      end
+    end
+
+    def applicable_schools
+      if ENV['DELETE_INACTIVE_USERS_AFTER'].present? && ENV['DELETE_INACTIVE_USERS_AFTER'].to_i != 0
+        School.all
+      else
+        School.where("configuration ? :key", :key => 'deleteInactiveUsersAfter')
+      end
+    end
+
+    def users_to_notify_deletion
+      User.where(school: applicable_schools).includes(:school).map do |user|
+
+        next if user.last_sign_in_at.blank? || (Time.zone.now - user.last_sign_in_at) < (inactivity_duration_for_user_deletion(user).months - 1.month)
+
+        next if user.account_expiry_notification_sent_at.present? && (user.account_expiry_notification_sent_at > inactivity_duration_for_user_deletion(user).months.ago)
+
+        user
+      end.compact
+    end
```

Comment:
This seems to be iterating over every user in applicable schools. This could be a huge result-set.

Instead, how about:

1. iterating over each school and collecting only those users who are applicable:
2. `nil`-ng the _notification sent at_ when a user signs in - I think this would allow the query to be simpler.

```suggestion
    def users_to_notify_deletion
      applicable_schools.map do |school|
        notify_after = notification_period(school)

        school.users
          .where('last_sign_in_at < ?', notify_after.months)
          .where.not(account_deletion_notification_sent_at: nil)
          .pluck(:id)
        end
      end.flatten
    end
```